openmdao/components/tests/test_exec_comp.py

```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
+            comp.declare_partials('*', ['x', 'z'], method='fd')
+            p.setup()
+            p['comp.x'] = 3.
+            p['comp.y'] = 4.
+            p['comp.z'] = 5.
+            p.run_model()
+            assert_near_equal(p['comp.out1'], np.ones(size) * 45., 1e-6)
+            assert_near_equal(p['comp.out2'], np.ones(size) * 21., 1e-6)
+
+            data = p.check_partials(out_stream=None)
+            self.assertEqual(list(data), ['comp'])
+
+    def test_register_check_partials_not_safe_mult_expr_err(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('unsafe', lambda x: x**2, complex_safe=False)
+            om.ExecComp.register('safe', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            comp = p.model.add_subsystem('comp', om.ExecComp(['out1 = unsafe(x) * z',
+                                                              'out2 = safe(y) + z'], shape=size))
+            comp.declare_partials('*', ['x'], method='fd')
+            p.setup()
+            p['comp.x'] = 3.
+            p['comp.y'] = 4.
+            p['comp.z'] = 5.
+            p.run_model()
+            assert_near_equal(p['comp.out1'], np.ones(size) * 45., 1e-6)
+            assert_near_equal(p['comp.out2'], np.ones(size) * 21., 1e-6)
+
+            with self.assertRaises(Exception) as cm:
+                data = p.check_partials(out_stream=None)
+            self.assertEquals(cm.exception.args[0],
+                              "'comp' <class ExecComp>: expression contains functions ['unsafe'] that are not complex safe. To fix this, call declare_partials('*', ['z'], method='fd') on this component prior to setup.")
+
+    def test_register_check_partials_safe(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+
+            data = p.check_partials(out_stream=None)
+            self.assertEqual(list(data), [])
+
+    def test_register_simple_arr_manual_partials_cs(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            comp = p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            comp.declare_partials('area_square', 'x', method='cs')
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+            J = p.compute_totals(of=['comp.area_square'], wrt=['comp.x'])
+            assert_near_equal(J['comp.area_square', 'comp.x'], np.eye(size) * 6., 1e-6)
```

Comment:
Might make sense to use a tighter tolerance on the 'cs' test, so you know it's not doing fd.

Reply:
done